crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -416,6 +424,22 @@ where
                     self.flow_merge(&pre_if);
                 }
             }
+            ast::Stmt::While(node) => {
+                self.visit_expr(&node.test);
+                let pre_loop = self.flow_snapshot();
+                let mut break_states: Vec<FlowSnapshot> = vec![];
+                std::mem::swap(&mut self.loop_break_states, &mut break_states);
+                self.visit_body(&node.body);
+                std::mem::swap(&mut self.loop_break_states, &mut break_states);
+                self.flow_merge(&pre_loop);
+                self.visit_body(&node.orelse);
+                for break_state in break_states {
+                    self.flow_merge(&break_state);
+                }
```

Comment:
The break state logic is a bit confusing. I think that's to some extend because we reuse the same name
```suggestion
								// Saved break states from any outer loop
                let mut saved_break_states = std::mem::take(&mut self.loop_break_states);
                self.visit_body(&node.body);
                let break_states = std::mem::replace(&mut self.loop_break_states, saved_break_states);
								for break_state in break_states {
                    self.flow_merge(&break_state);
                }

                self.flow_merge(&pre_loop);
                self.visit_body(&node.orelse);
```

I'm not sure if the ordering of the merges matters. If not, then I think having the merging of the break states (which are related to the for body) closer to `visit_body` helps understanding the flow.

Reply:
Ah, that is nicer, thanks!

The ordering of the merges does matter. The `else` block cannot see the break states, because breaking out of a loop bypasses the `else` clause.